src/json-ld.net/Core/DocumentLoader.cs

```diff
@@ -6,224 +6,108 @@
                     }
-                    string header = linkedContexts.First();
-                    string linkedUrl = header.Substring(1, header.IndexOf(">") - 1);
-                    string resolvedUrl = URL.Resolve(url, linkedUrl);
-                    var remoteContext = this.LoadDocument(resolvedUrl);
-                    doc.contextUrl = remoteContext.documentUrl;
-                    doc.context = remoteContext.document;
-                }
 
-                Stream stream = resp.GetResponseStream();
+                    var finalUrl = response.RequestMessage.RequestUri.ToString();
 
-                doc.DocumentUrl = req.Address.ToString();
-                doc.Document = JSONUtils.FromInputStream(stream);
-            }
-            catch (JsonLdError)
-            {
-                throw;
-            }
-            catch (WebException webException)
-            {
-                try
-                {
-                    resp = (HttpWebResponse)webException.Response;
-                    int baseStatusCode = (int)(Math.Floor((double)resp.StatusCode / 100)) * 100;
-                    if (baseStatusCode == 300)
+                    var contentType = GetJsonLDContentType(response.Content.Headers.ContentType.MediaType);
+
+                    if (contentType == JsonLDContentType.Other)
                     {
-                        string location = resp.Headers[HttpResponseHeader.Location];
-                        if (!string.IsNullOrWhiteSpace(location))
+                        throw new JsonLdError(JsonLdError.Error.LoadingDocumentFailed, url);
+                    }
+
+                    // For plain JSON, see if there's a context document linked in the HTTP response headers.
+                    if (contentType == JsonLDContentType.PlainJson && response.Headers.TryGetValues("Link", out var linkHeaders))
+                    {
+                        linkHeaders = linkHeaders.SelectMany((h) => h.Split(",".ToCharArray()))
+                                                    .Select(h => h.Trim()).ToArray();
+                        IEnumerable<string> linkedContexts = linkHeaders.Where(v => v.EndsWith("rel=\"http://www.w3.org/ns/json-ld#context\""));
+                        if (linkedContexts.Count() > 1)
                         {
-                            // TODO: Add recursion break or simply switch to HttpClient so we don't have to recurse on HTTP redirects.
-                            return LoadDocument(location);
+                            throw new JsonLdError(JsonLdError.Error.MultipleContextLinkHeaders);
                         }
+                        string header = linkedContexts.First();
+                        string linkedUrl = header.Substring(1, header.IndexOf(">") - 1);
+                        string resolvedUrl = URL.Resolve(finalUrl, linkedUrl);
+                        var remoteContext = this.LoadDocument(resolvedUrl);
```

Comment:
Invoking `LoadDocument()` syncrhonously inside `LoadDocumentAsync()` seems like a bad idea both because of `async` deadlock issues and because it will lead to infinite recursion given that `LoadDocument()` invokes `LoadDocumentAsync()`, no?

Reply:
Definitely a bad idea because `async`, but it shouldn't cause infinite recursion. `resolvedUrl` is a newly computed URL for a context document that got computed from a `Link` HTTP response header. Will fix this to be async.